google/gax/__init__.py

```diff
@@ -482,3 +496,180 @@ def __next__(self):
+                a client for the long-running operation service.
+            result_type (type): the class type of the result.
+            metadata_type (type, optional): the class type of the metadata.
+            call_options (google.gax.CallOptions, optional): the call options
+                that are used when reloading the operation.
+        """
+        self._operation = operation
+        self._client = client
+        self._result_type = result_type
+        self._metadata_type = metadata_type
+        self._call_options = call_options
+        self._done_clbks = []
+        self._process = None
+
+    def cancel(self):
+        """If last Operation's value of `done` is true, returns false;
+           otherwise, issues OperationsClient.cancel_operation and returns true.
+        """
+        if self.done():
+            return False
+
+        self._client.cancel_operation(self._operation.name)
+        return True
+
+    def result(self, timeout=None):
+        """Enters polling loop on OperationsClient.get_operation, and once
+           Operation.done is true, then returns Operation.response if successful
+           or throws GaxError if not successful.
+
+           This method will wait up to timeout seconds. If the call hasn't
+           completed in timeout seconds, then a RetryError will be raised.
+           timeout can be an int or float. If timeout is not specified or None,
+           there is no limit to the wait time.
+        """
+        if not self._poll(timeout).HasField('response'):
+            raise GaxError(self._operation.error.message)
+
+        return _from_any(self._result_type, self._operation.response)
+
+    def exception(self, timeout=None):
+        """Similar to result(), except returns the exception if any."""
+        if self._poll(timeout).HasField('error'):
+            return self._operation.error
+
+        return None
+
+    def cancelled(self):
+        """Return True if the call was successfully cancelled."""
+        self._get_operation()
+        return (self._operation.HasField('error') and
+                self._operation.error.code == code_pb2.CANCELLED)
+
+    def done(self):
+        """Issues OperationsClient.get_operation and returns value of
+           Operation.done.
+        """
+        return self._get_operation().done
+
+    def add_done_callback(self, fn):  # pylint: disable=invalid-name
+        """Enters a polling loop on OperationsClient.get_operation, and once the
+           operation is done or cancelled, calls the function with this
+           _OperationFuture. Added callables are called in the order that they
+           were added.
+        """
+        self._done_clbks.append(fn)
+        if self._process is None:
+            self._process = mp.Process(target=self._execute_clbks)
+            self._process.start()
+        elif not self._process.is_alive() and self._operation.done:
+            _try_callback(self, fn)
```

Comment:
I think it ends up looking messier because I have to check if the process is None twice:

```python
if self._process is not None and not self._process.is_alive() and self._operation.done:
    _try_callback(self, fn)
    return

self._done_clbks.append(fn)
if self._process is None:
    self._process = mp.Process(target=self._execute_clbks)
    self._process.start()
```